Better Tasks

Ryex opened the pull request:
<!--
Hey there! Thanks for your contribution.

Please make sure that your commits are signed off first.
If you don't know how that works, check out our contribution guidelines: https://github.com/PrismLauncher/PrismLauncher/blob/develop/CONTRIBUTING.md#signing-your-work
If you already created your commits, you can run `git rebase --signoff develop` to retroactively sign-off all your commits and `git push --force` to override what you have pushed already.

Note that signing and signing-off are two different things!
-->

Better Tasks:

- Tasks have UUIDs
  - easier to track in logs
  - can better track data related to them (like when propagating progress of child tasks)
- ProgressDialog has been revamped to report child task progress and status details (ie download speed)
- all existing tasks have been updated to propagate the progress of their child tasks
  - this mean download progress and speed is reported for effectively everything.
    - ~~or at least it would if QNetwork Replies reported total download size. for now only speed works.~~
    
 - add loading of logging rules from a `qtlogging.ini` file at the Prism data path root.
 an example file
 ```ini
[Rules]
# remove the debug lines, other log levels still get through
launcher.task.net.download.debug=false
# enable or disable whole catagories
launcher.task.net=true
launcher.task=false
launcher.task.net.upload=true
 ``` 

[bettertasks_pack_download.webm](https://user-images.githubusercontent.com/508861/229261221-c158c568-3e22-4cdf-9449-c989597fa602.webm)

[bettertasks_assets_download.webm](https://user-images.githubusercontent.com/508861/229261228-a5f788b8-7da4-42a1-bc0c-1c75de60b3a7.webm)

Ryex committed:
feat: better task tracking

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
feat: Propogated subtask progress

Oh boy this is big.

> TaskStepProgress struct is now QMetaObject compatabile and can be sent through signals
> Task now has a method to propogates sub task progress it must be signal bound by each task containing a task wishing to report progress of it's children.
> Downloads report speed
> Tasks now have UUIDS to track them
  - use when reporting
  - use when logging
  - use when storeing them or objects related to them

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Scrumplex commented on launcher/ui/dialogs/ProgressDialog.cpp:
```diff
@@ -1,29 +1,72 @@
+    auto type_range = type_max - type_min;
+    auto int_range = int_max - int_min;
+
+    return static_cast<int>((value - type_min) * int_range / type_range + int_min);
```
While this isn't a bad idea by itself, it doesn't work with the values we are dealing with here.

`int_range / type_range` evaluates to `2147483647 / 9223372036854775807 ≈ 2.328306435E−10`

Note that in this case it actually evaluates to `0`, as you are dividing two integers.

In any case, even if we would have a float here, every input value below `4294967298` would return less than `1`


To be honest, I think we can just silently ignore values larger than `int_max`, as we would be talking about file sizes > 2 GB, which should be fairly rare. Maybe just emit a warning in the log and continue.

Scrumplex replied:
(if we just ignore this map function and pass the progress values directly below, the progress bars work!)

Ryex replied:
Doy, I should have known this would be a problem. I'll need to think of a better way to map the 64bit int range progress we get  into to a percentage. In fact that is likely exactly what I should be doing. Map to a floating point percent and then map that to a reasonable into range.

Scrumplex replied:
alternatively you could also do `current / total * 100` and pass that as the current value, while passing `100` as the new total. Or swap out 100 with a constant that defines precision.

Note that my example here is integer division, you might want to work with floats instead

Ryex replied:
that's exactly what the fixed function does, except it used _INT_MAX instead of 100 (more precision for large downloads)

Scrumplex commented on launcher/net/Download.cpp:
```diff
@@ -284,11 +331,11 @@ void Download::downloadReadyRead()
+        qCCritical(DownloadLogC) << getUid().toString() << "Cannot write download data! illegal status " << m_status;
     }
 }
 
```
Consider adding all of these changes to Upload.cpp as well :D

Ryex replied:
It's in the plan, yes.

Scrumplex commented on launcher/ui/dialogs/ProgressDialog.cpp:
```diff
@@ -54,10 +97,17 @@ ProgressDialog::~ProgressDialog()
 }
 
 void ProgressDialog::updateSize()
-{
+{   
+    QSize lastSize = this->size();
     QSize qSize = QSize(480, minimumSizeHint().height());
     resize(qSize);
     setFixedSize(qSize);
+    // keep the dialog in the center after a resize
+    if (lastSize != qSize)
+        this->move(
+            this->parentWidget()->x() + (this->parentWidget()->width() - this->width()) / 2,
+            this->parentWidget()->y() + (this->parentWidget()->height() - this->height()) / 2
+        );
 }
```
I don't really think we should be doing this at all? Like, the window should just have some minimum sizes, depending on what is shown. But we probably shouldn't prevent the user from resizing this window, especially if there are multiple tasks visible in the dialog.

Ryex replied:
You would think that that would work and itd hoped it would but the window layout flat out breaks after unhideing the scroll area unless we do a manual resize.

Ryex replied:
NVM I fixed it, the user can now resize the dialog the function now sets the minimum size and recenters only after unhideing

Scrumplex commented:
(apart from my comments above, excellent work so far!)

Ryex committed:
fix: properly map progress range
- doument PCRE used for URL compacting

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
feat: add `details` signal to `Task`
feat: add details to mod pack downloading
feat: add logging rule sloading form `ligging.ini at data path root
feat: add `launcher.task` `launcher.task.net` and `launcher.task.net.[down|up]load` logging categories
fix: add new subtask progress to the end of the lay out not the beginning (cuts down on flickering)

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex commented:
ok, ready for a more in depth review. I feel good about what I've achieved with this in terms of QOL changes.

Ryex commented:
ugg, why are some of the QT6 task test's failing....

Scrumplex commented on launcher/Application.cpp:
```diff
@@ -410,6 +411,24 @@ Application::Application(int &argc, char **argv) : QApplication(argc, argv)
+                    qDebug() << "    " << rule;
+                }
+                auto rules_str = rules.join("\n");
+                QLoggingCategory::setFilterRules(rules_str);
```
Why don't we just use Qt's built-in support for this (using the `QT_LOGGING_CONF` env var)


See https://doc.qt.io/qt-6/qloggingcategory.html#configuring-categories

Ryex replied:
Because of the rule load order precedence. Rules through the `setFilterRules` Function can be overridden by a file specified by a `QT_LOGGING_CONF`
Which can in turn be overridden by raw rules specified in `QT_LOGGING_RULES`. doing it this way still allows for user overrides at every launch.

Also, this way we can potentially distribute a default rules file so that not every debug line spams the logs. We would just have to add a an additional check for the file next to the binary or usr/share or where ever for the default rules to load

flowln replied:
I think this file should be identical to `qtlogging.ini`, including the name and the format. I also think this `INIFile` is trash, [`QSettings`](https://doc.qt.io/qt-5/qsettings.html) has support for INI files and works way better, so you can parse the `[Rules]` section with it.

Ryex replied:
hmm, yeah. I'll move it to use `QSettings`. I'll set up a default rules file.

Ryex replied:
ok, now it uses QSettings (supports and uses the `[Rules]` section) and checks in order:
the datapath, the `AppDataLocation` via `QStandardPaths`, next to the binary.

I added a default `qtloggining.ini` and set it up to be installed. but that part will need to be vetted that it finds the file correctly on windows and mac.

Ryex committed:
feat: add download size to download details alongside speed

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
feat: add UID to debug lines of upload tasks

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

flowln commented on launcher/net/HttpMetaCache.cpp:
```diff
@@ -124,7 +126,7 @@ auto HttpMetaCache::resolveEntry(QString base, QString resource_path, QString ex
     auto current_time = QDateTime::currentSecsSinceEpoch();
     if (entry->isExpired(current_time - ( file_last_changed / 1000 ))) {
-        qWarning() << "Removing cache entry because of old age!";
+        qCWarning(taskNetLogC) << "[HttpMetaCache]" << "Removing cache entry because of old age!";
```
why not create a subcategory of the logging category for the HttpMetaCache?

flowln commented on launcher/net/logging.h:
`logging.h` -> `Logging.h`
our files generally begin with capital letter, and you can specify `#include "net/Logging.h"` in the files that use it, so as to not conflict with the existing `Logging.h`.

flowln commented on launcher/tasks/Task.h:
```diff
@@ -36,9 +37,37 @@
+enum class TaskStepState {
+    Waiting,
+    Running,
+    Failed,
+    Succeeded,
+    Finished
```
why is `Finished` a separate state? wouldn't a task be finished when state is either `Failed` or `Succeeded`?

Ryex replied:
I'll remove it. I was thinking too  far ahead

flowln commented on launcher/ui/dialogs/ProgressDialog.cpp:
```diff
@@ -1,29 +1,69 @@
+
+// map a value in a numaric range of an arbatray type to between 0 and INT_MAX
+// for getting the best percision out of the qt progress bar
+template<typename T, typename = typename std::enable_if<std::is_arithmetic<T>::value, T>::type> 
```
```suggestion
template<typename T, std::enable_if_t<std::is_arithmetic_v<T>, bool> = true> 
```

flowln commented on launcher/tasks/ConcurrentTask.cpp:
```diff
@@ -138,28 +138,89 @@ void ConcurrentTask::subTaskFailed(Task::Ptr task, const QString& msg)
+
+void ConcurrentTask::updateStepProgress()
+{
+   qint64 current = 0, total = 0;
+   for ( auto taskProgress : m_task_progress ) {
+       current += taskProgress->current;
+       total += taskProgress->total;
+   }
+
+   m_stepProgress = current;
+   m_stepTotalProgress = total;
+   emit stepProgress(m_task_progress.values());
```
firstly, this needs a clang-format

secondly, this function is the main culprit of the failing tests. Your changes made tasks much more heavyweight in processing time, and this function (particularly the `.values()` call) dominates the test time: 
![kcachegrind view of the Task test running in this PR, highlighting the impact of updateStepProgress()](https://user-images.githubusercontent.com/9145768/229360644-1cf9b4e7-a0b8-4822-9742-430621309548.png)

Ryex replied:
hmm, yeah. propogating the whole list every time was clearly a bad move. I was initialy thinking I could reduce the the number of signal chain calls if I did it that way but that never happened.

I'll try to make it work passing one task at a time the struct copy constructor has to be faster than the `QMap::values` call.

Ryex replied:
> firstly, this needs a clang-format
> 
> secondly, this function is the main culprit of the failing tests. Your changes made tasks much more heavyweight in processing time, and this function (particularly the `.values()` call) dominates the test time: ![kcachegrind view of the Task test running in this PR, highlighting the impact of updateStepProgress()](https://user-images.githubusercontent.com/9145768/229360644-1cf9b4e7-a0b8-4822-9742-430621309548.png)


OK, clearly I need to do my own testing with kcachegrind to figure this out. the task test still fails

flowln replied:
Here's a basic patch fixing most of the performance issue with this part of the code (maybe there's stuff to improve here, which is why i didn't add it directly as a commit. it's just a quick fix based on the output of kcachegrind :P). The main culprit here is the fact that you iterate over all subtask progresses every time you update the task's step progress in `updateStepProgress`, so it ends up iterating over the number of all tasks every time a task gets update, and the QHash `iterator++` method gets called a bunch of times, which adds up to a lot of time.

<details>
<summary>Patch</summary>

```patch
diff --git a/launcher/tasks/ConcurrentTask.cpp b/launcher/tasks/ConcurrentTask.cpp
index 8434ecdc..417b0538 100644
--- a/launcher/tasks/ConcurrentTask.cpp
+++ b/launcher/tasks/ConcurrentTask.cpp
@@ -138,10 +138,11 @@ void ConcurrentTask::startNext()
     connect(next.get(), &Task::progress, this, [this, next](qint64 current, qint64 total) { subTaskProgress(next, current, total); });
 
     m_doing.insert(next.get(), next);
-    m_task_progress.insert(next->getUid(), std::make_shared<TaskStepProgress>(TaskStepProgress({ next->getUid() })));
+    auto task_progress = std::make_shared<TaskStepProgress>(next->getUid());
+    m_task_progress.insert(next->getUid(), task_progress);
 
     updateState();
-    updateStepProgress();
+    updateStepProgress(*task_progress.get(), Operation::ADDED);
 
     QCoreApplication::processEvents();
 
@@ -166,7 +167,7 @@ void ConcurrentTask::subTaskSucceeded(Task::Ptr task)
 
     emit stepProgress(*task_progress.get());
     updateState();
-    updateStepProgress();
+    updateStepProgress(*task_progress.get(), Operation::REMOVED);
     startNext();
 }
 
@@ -184,7 +185,7 @@ void ConcurrentTask::subTaskFailed(Task::Ptr task, const QString& msg)
 
     emit stepProgress(*task_progress.get());
     updateState();
-    updateStepProgress();
+    updateStepProgress(*task_progress.get(), Operation::REMOVED);
     startNext();
 }
 
@@ -193,9 +194,8 @@ void ConcurrentTask::subTaskStatus(Task::Ptr task, const QString& msg)
     auto task_progress = m_task_progress.value(task->getUid());
     task_progress->status = msg;
     task_progress->state = TaskStepState::Running;
-    
+
     emit stepProgress(*task_progress.get());
-    updateStepProgress();
 }
 
 void ConcurrentTask::subTaskDetails(Task::Ptr task, const QString& msg)
@@ -203,50 +203,69 @@ void ConcurrentTask::subTaskDetails(Task::Ptr task, const QString& msg)
     auto task_progress = m_task_progress.value(task->getUid());
     task_progress->details = msg;
     task_progress->state = TaskStepState::Running;
-    
+
     emit stepProgress(*task_progress.get());
-    updateStepProgress();
 }
 
 void ConcurrentTask::subTaskProgress(Task::Ptr task, qint64 current, qint64 total)
 {
     auto task_progress = m_task_progress.value(task->getUid());
 
+    task_progress->old_current = task_progress->current;
+    task_progress->old_total = task_progress->total;
+
     task_progress->current = current;
     task_progress->total = total;
     task_progress->state = TaskStepState::Running;
-    
+
     emit stepProgress(*task_progress.get());
-    updateStepProgress();
+    updateStepProgress(*task_progress.get(), Operation::CHANGED);
     updateState();
 }
 
 void ConcurrentTask::subTaskStepProgress(Task::Ptr task, TaskStepProgress task_progress)
 {
+    Operation op = Operation::ADDED;
+
     if (!m_task_progress.contains(task_progress.uid)) {
         m_task_progress.insert(task_progress.uid, std::make_shared<TaskStepProgress>(task_progress));
+        op = Operation::ADDED;
     } else {
         auto tp = m_task_progress.value(task_progress.uid);
+
+        tp->old_current = tp->current;
+        tp->old_total = tp->total;
+
         tp->current = task_progress.current;
         tp->total = task_progress.total;
         tp->status = task_progress.status;
         tp->details = task_progress.details;
+
+        op = Operation::CHANGED;
     }
-    
+
     emit stepProgress(task_progress);
-    updateStepProgress();
+    updateStepProgress(task_progress, op);
 }
 
-void ConcurrentTask::updateStepProgress()
+void ConcurrentTask::updateStepProgress(TaskStepProgress const& changed_progress, Operation op)
 {
-    qint64 current = 0, total = 0;
-    for (auto taskProgress : m_task_progress) {
-        current += taskProgress->current;
-        total += taskProgress->total;
+    switch(op) {
+        case Operation::ADDED:
+            m_stepProgress += changed_progress.current;
+            m_stepTotalProgress += changed_progress.total;
+            break;
+        case Operation::REMOVED:
+            m_stepProgress -= changed_progress.current;
+            m_stepTotalProgress -= changed_progress.total;
+            break;
+        case Operation::CHANGED:
+            m_stepProgress -= changed_progress.old_current;
+            m_stepTotalProgress -= changed_progress.old_total;
+            m_stepProgress += changed_progress.current;
+            m_stepTotalProgress += changed_progress.total;
+            break;
     }
-
-    m_stepProgress = current;
-    m_stepTotalProgress = total;
 }
 
 void ConcurrentTask::updateState()
diff --git a/launcher/tasks/ConcurrentTask.h b/launcher/tasks/ConcurrentTask.h
index 974e8d4d..284fa345 100644
--- a/launcher/tasks/ConcurrentTask.h
+++ b/launcher/tasks/ConcurrentTask.h
@@ -83,7 +83,8 @@ protected:
     // NOTE: This is not thread-safe.
     [[nodiscard]] unsigned int totalSize() const { return m_queue.size() + m_doing.size() + m_done.size(); }
 
-    void updateStepProgress();
+    enum class Operation { ADDED, REMOVED, CHANGED };
+    void updateStepProgress(TaskStepProgress const& changed_progress, Operation);
 
     virtual void updateState();
 
diff --git a/launcher/tasks/Task.h b/launcher/tasks/Task.h
index 9ae70270..9b40610c 100644
--- a/launcher/tasks/Task.h
+++ b/launcher/tasks/Task.h
@@ -54,9 +54,15 @@ enum class TaskStepState {
 Q_DECLARE_METATYPE(TaskStepState)
 
 struct TaskStepProgress {
+    TaskStepProgress(QUuid uid) : uid(uid) {}
+
     QUuid uid; 
     qint64 current = 0;
     qint64 total = -1;
+
+    qint64 old_current = 0;
+    qint64 old_total = -1;
+
     QString status = "";
     QString details = "";
     TaskStepState state = TaskStepState::Waiting;

``` 

</details>

Ryex committed:
fix: clean up license headers for Tasks

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
refactor: logging.h -> Logging.h

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
refactor: propogate only only one StepProgress at a time

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
feat: default qtlogging.ini file

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
fix: no need to loop all sub tasks

pathc by flowin

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>
Signed-off-by: flow <flowlnlnln@gmail.com>

Ryex committed:
fix: no need for const bool

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex commented on launcher/tasks/Task.cpp:
```diff
@@ -135,16 +151,21 @@
     emit finished();
 }
 
+void Task::propogateStepProgress(TaskStepProgress task_progress)
```
oh, I wish I could Code QL, but it needs to serialize over the signals interface.

Ryex replied:
... or not.

Apparently Qt handles this by making copies of refs as needed to ensure it's garentees like if it needs to cross a thread boundary.

Ryex committed:
refactor: Qt can handle const& in signals and slots

While most Qt types cna use implicit data sharing
pasing our own structs means copies. const& ensure
it's only copied as needed by Qt.

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

DioEgizio commented:
<img width="1254" alt="huh wtf" src="https://user-images.githubusercontent.com/83089242/230711073-80627f24-a48d-4a6c-9458-261ff99faf82.png">
there's something wrong in this pr (doesn't happen in develop)

DioEgizio commented:
https://github.com/PrismLauncher/PrismLauncher/pull/961/commits/d4297b75776e5f503bffbca345d7b3405aaeb76d is the last commit where this doesn't happen, i suppose it's caused by that ini file

Ryex commented:
> [d4297b7](https://github.com/PrismLauncher/PrismLauncher/commit/d4297b75776e5f503bffbca345d7b3405aaeb76d) is the last commit where this doesn't happen, i suppose it's caused by that ini file

What the... This is MacOS?

DioEgizio commented:
I'm on macOS ye

Ryex commented:
time to load up my VM then....

Ryex commented:
> <img alt="huh wtf" width="1254" src="https://user-images.githubusercontent.com/83089242/230711073-80627f24-a48d-4a6c-9458-261ff99faf82.png">
> there's something wrong in this pr (doesn't happen in develop)

@DioEgizio  I can't seem to replicate this on my VM. what configure args are you passing. does it actually load the `qtlogging.ini` file? (should be at the beginning of the logs)

DioEgizio commented:
Did you try reproducing it building yourself or did you try the [CI builds](https://github.com/PrismLauncher/PrismLauncher/actions/runs/4642987886)?

anyway there's a possibility the latest develop commit fixed this, so try just rebasing

Ryex commented:
> Did you try reproducing it building yourself or did you try the [CI builds](https://github.com/PrismLauncher/PrismLauncher/actions/runs/4642987886)?
> 
> anyway there's a possibility the latest develop commit fixed this, so try just rebasing. 

I was building myself with QT 6.5. hard to debug a ci-build. I'll try one now though.

Ryex commented:
ok, yes, the CI builds do have it somewhat
![image](https://user-images.githubusercontent.com/508861/230736117-7536445d-d845-46a7-9114-1b0999646abb.png)

Ryex committed:
fix: prevent logspam, fix MacOS theme artifacts

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex commented:
For the record enabling the logging category `qt.qpa.drawing.debug` causes theme artifacts on MacOS

Ryex committed:
feat: estimate remining time on downloads

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

DioEgizio commented:
forgor to add but now everything works fine on macos

DioEgizio commented:
Conflicting with develop

Ryex commented:
updated

TayouVR commented:
I think this was finished, right?
Just needs one more approval.
I'm not the most qualified, but in my eyes the code looks good.

Ryex commented:
yep, this is finished

flowln commented on launcher/Application.cpp:
```diff
@@ -410,6 +412,47 @@ Application::Application(int &argc, char **argv) : QApplication(argc, argv)
+
+        // search the dataPath()
+        
+        if(!foundLoggingRules && ! isPortable()) {
```
I think this should also not apply when using the `-d` flag (i.e. it should only search the AppDataLocation when the app is using the permanent data path)

Ryex replied:
good point. added a check for `&& dirParam.isEmpty()`.

flowln commented on launcher/java/JavaInstallList.cpp:
```diff
@@ -170,6 +170,7 @@ void JavaListLoadTask::executeTask()
     m_job.reset(new JavaCheckerJob("Java detection"));
     connect(m_job.get(), &Task::finished, this, &JavaListLoadTask::javaCheckerFinished);
     connect(m_job.get(), &Task::progress, this, &Task::setProgress);
+    // stepProgress?
```
:thinking:

Ryex replied:
```suggestion

```

A marker for me to check if the `stepProgress` signal was needed here

flowln commented on launcher/net/Download.cpp:
```diff
@@ -140,13 +242,35 @@ void Download::executeTask()
+
+    // use milliseconds for speed precision
+    auto elapsed_ms = std::chrono::duration_cast<std::chrono::milliseconds>(elapsed);
+    auto bytes_recived_since = bytesReceived - m_last_progress_bytes;
+    auto dl_speed_bps = (double)bytes_recived_since / elapsed_ms.count() * 1000;
```
```suggestion
    auto bytes_received_since = bytesReceived - m_last_progress_bytes;
    auto dl_speed_bps = (double)bytes_received_since / elapsed_ms.count() * 1000;
```

flowln commented on launcher/net/Download.cpp:
```diff
@@ -140,13 +242,35 @@ void Download::executeTask()
+    auto dl_speed_bps = (double)bytes_recived_since / elapsed_ms.count() * 1000;
+    auto remaing_time_s = (bytesTotal - bytesReceived) / dl_speed_bps;
+
+    // current bytes out of total bytes
```
```suggestion
    //: Current amount of bytes downloaded, out of the total amount of bytes in the download
```

flowln commented on launcher/net/Download.cpp:
```diff
@@ -140,13 +242,35 @@ void Download::executeTask()
+    
+    QString dl_speed_str;
+    if (elapsed_ms.count() > 0) {
+        // bytes per second
```
```suggestion
        //: Download speed, in bytes per second (remaining download time in parenthesis)
```

flowln commented on launcher/net/Download.cpp:
```diff
@@ -140,13 +242,35 @@ void Download::executeTask()
+        // bytes per second
+        dl_speed_str = tr("%1/s (%2)").arg(humanReadableFileSize(dl_speed_bps)).arg(humanReadableDuration(remaing_time_s));
+    } else {
+        dl_speed_str = tr("0 b/s");
```
```suggestion
		//: Download speed at 0 bytes per second
        dl_speed_str = tr("0 B/s");
```

flowln commented on launcher/net/Download.cpp:
```diff
@@ -140,13 +242,35 @@ void Download::executeTask()
+    QString dl_speed_str;
+    if (elapsed_ms.count() > 0) {
+        // bytes per second
+        dl_speed_str = tr("%1/s (%2)").arg(humanReadableFileSize(dl_speed_bps)).arg(humanReadableDuration(remaing_time_s));
```
```suggestion
        dl_speed_str = tr("%1 /s (%2)").arg(humanReadableFileSize(dl_speed_bps)).arg(humanReadableDuration(remaing_time_s));
```

flowln commented on launcher/net/NetAction.h:
```diff
@@ -35,18 +36,38 @@
 #include "QObjectPtr.h"
 #include "tasks/Task.h"
 
+static const QStringList s_units_si  {"kb", "MB", "GB", "TB"};
+static const QStringList s_units_kibi {"kiB", "MiB", "Gib", "TiB"};
```
```suggestion
static const QStringList s_units_si  {"kB", "MB", "GB", "TB"};
static const QStringList s_units_kibi {"KiB", "MiB", "Gib", "TiB"};
```

<details>
<summary>kibi</summary>

![kibe](https://user-images.githubusercontent.com/9145768/236562733-5cf346d3-adce-4144-bcdf-bc9b8ecd6a54.png)

</details>

Ryex replied:
I know there is a more technical name for the unit system but...  ~kibi bytes~

flowln commented on launcher/tasks/ConcurrentTask.cpp:
```diff
@@ -32,11 +65,9 @@ void ConcurrentTask::addTask(Task::Ptr task)
-    for (int i = 0; i < num_starts; i++) {
-        QMetaObject::invokeMethod(this, &ConcurrentTask::startNext, Qt::QueuedConnection);
-    }
+    // Start One task, startNext hadels starting the up to the m_total_max_size
```
```suggestion
    // Start one task, startNext handles starting the up to the m_total_max_size
```

flowln commented on launcher/tasks/ConcurrentTask.cpp:
```diff
@@ -139,27 +179,128 @@ void ConcurrentTask::subTaskFailed(Task::Ptr task, const QString& msg)
+                      .arg(QString::number(m_doing.count()), QString::number(m_done.count()), QString::number(totalSize())));
+    } else {
+        setProgress(m_stepProgress, m_stepTotalProgress);
+        QString status = tr("Please wait ...");
+        if (m_queue.size() > 0) {
+            status = tr("Waiting for 1 task to start ...");
```
```suggestion
        QString status = tr("Please wait...");
        if (m_queue.size() > 0) {
            status = tr("Waiting for a task to start...");
```

flowln commented on launcher/tasks/Task.h:
```diff
@@ -94,7 +129,8 @@ class Task : public QObject, public QRunnable {
     void status(QString status);
-    void stepStatus(QString status);
+    void details(QString details);
+    void stepProgress(TaskStepProgress const& task_progress); // 
```
```suggestion
    void stepProgress(TaskStepProgress const& task_progress);
```

flowln commented on launcher/ui/dialogs/ProgressDialog.cpp:
```diff
@@ -1,29 +1,69 @@
+#include "ui/widgets/SubTaskProgressBar.h"
+
+
+// map a value in a numaric range of an arbatray type to between 0 and INT_MAX
+// for getting the best precision out of the qt progress bar
```
```suggestion
// map a value in a numeric range of an arbitrary type to between 0 and INT_MAX
// for getting the best precision out of the qt progress bar
```

does it even make sense though? how are you getting more precision from the thing, the only thing I see happening is double precision adding a bit of variability to the end value, but that's not precision, that's just noise o.O

Ryex replied:
Technically this could be accomplished by mapping to a much smaller range because the smallest unit it can draw is one pixel width and the max number of those across a screen is going to be much less than 4000

the issue is that normal type coercion will just lose data with types larger than `INT_MAX`.
Also, for the progress bar it's somewhat important to use a consistent range size so that it doesn't jump around, and looks smooth as it progresses next to other bars.

flowln commented on launcher/ui/widgets/SubTaskProgressBar.h:
```diff
@@ -0,0 +1,50 @@
+ */
+#pragma once
+
+#include <qobjectdefs.h>
+#include <qwidget.h>
```
```suggestion
```

flowln reviewed (approved):
lgtm! _sorry for the thousand comments on consecutive lines, github should allow me to merge those, but it doesn't :pensive:_

Ryex committed:
Apply suggestions from code review

Co-authored-by: flow <flowlnlnln@gmail.com>
Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
change: don't search appdata locaiton for logging rules if using custom data dir

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Scrumplex commented on launcher/Application.cpp:
```diff
@@ -410,6 +413,47 @@ Application::Application(int &argc, char **argv) : QApplication(argc, argv)
+            }
+        }
+
+        if(!QFile::exists(logRulesPath)) {
```
From what I can tell, foundLoggingRules can only be true if a file has already been found. So I guess we can just use that?

```suggestion
        if(!foundLoggingRules) {
```

Scrumplex commented on launcher/Application.cpp:
```diff
@@ -410,6 +413,47 @@ Application::Application(int &argc, char **argv) : QApplication(argc, argv)
+        // search the dataPath()
+        
+        if(!foundLoggingRules && !isPortable() && dirParam.isEmpty()) {
+            logRulesPath = QStandardPaths::locate(QStandardPaths::AppDataLocation, logRulesFile);
```
```suggestion
            logRulesPath = QStandardPaths::locate(QStandardPaths::AppDataLocation, "../" + logRulesFile);
```

Note that `QStandardPaths::AppDataLocation` will include the Qt organization name in the path (`PrismLauncher/PrismLauncher`), so we need to go up a level. Though I am not sure if this works as expected.

Ryex replied:
it certainly appears to (I've tested it locally)

Scrumplex commented on launcher/modplatform/atlauncher/ATLPackInstallTask.cpp:
```diff
@@ -845,10 +846,12 @@ void PackInstallTask::downloadMods()
     });
     connect(jobPtr.get(), &NetJob::progress, [&](qint64 current, qint64 total)
-    {
+    {   
```
trailing spaces

Scrumplex commented on launcher/net/Download.cpp:
```diff
@@ -48,12 +51,106 @@
+    std_duration -= m;
+    auto s = std::chrono::duration_cast<std::chrono::seconds>(std_duration);
+    std_duration -= s;
+    auto ms = std::chrono::duration_cast<std::chrono::milliseconds>(std_duration);
+
+    auto dc = d.count();
+    auto hc = h.count();
+    auto mc = m.count();
+    auto sc = s.count();
+    auto msc = ms.count();
+
+    if (dc) {
+        os << dc << "days";
+    }
+    if (hc) {
+        if (dc)
+            os << " ";
+        os << qSetFieldWidth(2) << hc << "h";
+    }
+    if (mc) {
+        if (dc || hc)
+            os << " ";
+        os << qSetFieldWidth(2) << mc << "m";
+    }
+    if (dc || hc || mc || sc) {
+        if (dc || hc || mc)
+            os << " ";
+        os << qSetFieldWidth(2) << sc << "s";
+    }
+    if ((msc && (precision > 0)) || !(dc || hc || mc || sc)) {
+        if (dc || hc || mc || sc)
+            os << " ";
+        os << qSetFieldWidth(0) << qSetRealNumberPrecision(precision) << msc << "ms";
+    }
+
+    os.flush();
+
+    return outStr;
+}
+
```
This should probably go into `launcher/MMCTime.cpp`.

Also, having some tests for this code would be nice, but not a requirement for merging.

Scrumplex commented on launcher/net/Download.cpp:
```diff
@@ -48,12 +51,106 @@
+    if (str_url.length() <= max_len)
+        return str_url;
+
+    /* this is a PCRE regular expression that splits a URL (given by the display rules above) into 5 capture groups
+     * the scheme (ie https://) is group 1
+     * the host (with trailing /) is group 2
+     * the first part of the path (with trailing /) is group 3
+     * the last part of the path (with leading /) is group 5
+     * the remainder of the URL is in the .* and in group 4
+     *
+     * See: https://regex101.com/r/inHkek/1
+     * for an interactive breakdown
+     */ 
+    QRegularExpression re(R"(^([\w]+:\/\/)([\w._-]+\/)([\w._-]+\/)(.*)(\/[^]+[^]+)$)");
```
We could also extract some of these using QUrl's own parser. And as URLs can be very complex, I would prefer if we relied on a proper parser than using RegEx.

Ryex replied:
I actually tried using qurl first but I ended up needing most of this regular expression anyway to split the path. Adding a portion to capture the scheme and host simplifyed the code significantly. 

The point isn't to parse the parts of a url but to place an elipsus in the middle of the path intelagetly to truncate it's display. For that purpose I feel this approach with the regular expression is quite suitable seeing as there is no need to properly parse port or user athority or query fragments. (these are explicitly stripped out beforehand)

Ryex replied:
That said I think I came up with a non regex version.

Scrumplex commented on launcher/net/Download.cpp:
```diff
@@ -123,9 +222,12 @@ void Download::executeTask()
 
     QNetworkReply* rep = m_network->get(request);
-
+    
```
spaces

Scrumplex commented on launcher/net/Download.cpp:
```diff
@@ -140,13 +242,36 @@ void Download::executeTask()
+        //: Download speed, in bytes per second (remaining download time in parenthesis)
+        dl_speed_str = tr("%1 /s (%2)").arg(humanReadableFileSize(dl_speed_bps)).arg(humanReadableDuration(remaing_time_s));
+    } else {
+		//: Download speed at 0 bytes per second
```
tabs to spaces

Scrumplex commented on launcher/CMakeLists.txt:
```diff
@@ -563,6 +565,55 @@ ecm_qt_declare_logging_category(CORE_SOURCES
+    IDENTIFIER taskNetLogC
+    CATEGORY_NAME "launcher.task.net"
+    DEFAULT_SEVERITY Debug
+    DESCRIPTION "task network action"
+    EXPORT "${Launcher_Name}"
+)
+
+ecm_qt_export_logging_category(
+    IDENTIFIER taskDownloadLogC
+    CATEGORY_NAME "launcher.task.net.download"
+    DEFAULT_SEVERITY Debug
+    DESCRIPTION "task network download actions"
+    EXPORT "${Launcher_Name}"
+)
+ecm_qt_export_logging_category(
+    IDENTIFIER taskUploadLogC
+    CATEGORY_NAME "launcher.task.net.upload"
+    DEFAULT_SEVERITY Debug
+    DESCRIPTION "task network upload actions"
+    EXPORT "${Launcher_Name}"
+)
+
+ecm_qt_export_logging_category(
+    IDENTIFIER taskMetaCacheLogC
+    CATEGORY_NAME "launcher.task.net.metacache"
+    DEFAULT_SEVERITY Debug
+    DESCRIPTION "task network meta-cache actions"
+    EXPORT "${Launcher_Name}"
+)
+
+ecm_qt_export_logging_category(
+    IDENTIFIER taskHttpMetaCacheLogC
+    CATEGORY_NAME "launcher.task.net.metacache.http"
+    DEFAULT_SEVERITY Debug
+    DESCRIPTION "task network http meta-cache actions"
+    EXPORT "${Launcher_Name}"
+)
+
+
+
```
Any reason why you don't use `ecm_qt_declare_logging_category` like above?

Ryex replied:
the `ecm_qt_declare_logging_category` cmake function forces the creation of a new header file for each category.
I felt it preferable to manually create the logging catigories in one header and mark them for export here

Scrumplex replied:
I see. This is okay then :D

Scrumplex commented on launcher/net/Logging.cpp:
```diff
@@ -0,0 +1,26 @@
+// SPDX-License-Identifier: GPL-3.0-only
+/*
+ *  Prism Launcher - Minecraft Launcher
+ *  Copyright (C) 2023 Rachel Powers <508861+Ryex@users.noreply.github.com>
+ *
+ *  This program is free software: you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation, version 3.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ *
+ */
+
+#include "net/Logging.h"
+
+Q_LOGGING_CATEGORY(taskNetLogC, "launcher.task.net")
+Q_LOGGING_CATEGORY(taskDownloadLogC, "launcher.task.net.download")
+Q_LOGGING_CATEGORY(taskUploadLogC, "launcher.task.net.upload")
+Q_LOGGING_CATEGORY(taskMetaCacheLogC, "launcher.task.net.metacache")
+Q_LOGGING_CATEGORY(taskHttpMetaCacheLogC, "launcher.task.net.metacache.http")
```
we could get rid of this if we use `ecm_qt_declare_logging_category` in CMake

Scrumplex commented on launcher/net/Logging.h:
```diff
@@ -0,0 +1,28 @@
+// SPDX-License-Identifier: GPL-3.0-only
+/*
+ *  Prism Launcher - Minecraft Launcher
+ *  Copyright (C) 2023 Rachel Powers <508861+Ryex@users.noreply.github.com>
+ *
+ *  This program is free software: you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation, version 3.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ *
+ */
+ 
+#pragma  once
+
+#include <QLoggingCategory>
+
+Q_DECLARE_LOGGING_CATEGORY(taskNetLogC)
+Q_DECLARE_LOGGING_CATEGORY(taskDownloadLogC)
+Q_DECLARE_LOGGING_CATEGORY(taskUploadLogC)
+Q_DECLARE_LOGGING_CATEGORY(taskMetaCacheLogC)
+Q_DECLARE_LOGGING_CATEGORY(taskHttpMetaCacheLogC)
```
we could get rid of this if we use `ecm_qt_declare_logging_category` in CMake

Scrumplex commented on launcher/net/Download.cpp:
```diff
@@ -48,12 +51,106 @@
+
 namespace Net {
 
+QString truncateUrlHumanFriendly(QUrl &url, int max_len, bool hard_limit = false)
```
maybe move this into `launcher/StringUtils.cpp`?

Scrumplex commented on launcher/net/NetAction.h:
```diff
@@ -35,18 +36,38 @@
+static const QStringList s_units_si  {"kB", "MB", "GB", "TB"};
+static const QStringList s_units_kibi {"KiB", "MiB", "Gib", "TiB"};
+
+inline QString humanReadableFileSize(double bytes, bool use_si = false, int decimal_points = 1) {
```
maybe move this into `launcher/StringUtils.cpp`?

Scrumplex commented on launcher/net/NetAction.h:
```diff
@@ -35,18 +36,38 @@
 class NetAction : public Task {
     Q_OBJECT
-   protected:
+protected:
```
I am sure that clang-format would disagree with this change

Scrumplex commented on launcher/net/NetAction.h:
```diff
@@ -35,18 +36,38 @@
     explicit NetAction() : Task() {};
 
-   public:
+public:
```
ditto

Scrumplex commented on launcher/net/NetAction.h:
```diff
@@ -55,23 +76,23 @@ class NetAction : public Task {
     void setNetwork(shared_qobject_ptr<QNetworkAccessManager> network) { m_network = network; }
 
-   protected slots:
+protected slots:
```
ditto

Scrumplex commented on launcher/net/NetAction.h:
```diff
@@ -55,23 +76,23 @@ class NetAction : public Task {
     virtual void downloadReadyRead() = 0;
 
-   public slots:
+public slots:
```
ditto

Scrumplex commented on launcher/net/NetAction.h:
```diff
@@ -55,23 +76,23 @@ class NetAction : public Task {
     }
 
-   protected:
+protected:
```
ditto

Scrumplex commented on launcher/net/NetAction.h:
```diff
@@ -55,23 +76,23 @@ class NetAction : public Task {
     void executeTask() override {};
 
-   public:
+public:
```
ditto

Scrumplex commented on launcher/tasks/ConcurrentTask.h:
```diff
@@ -56,9 +94,12 @@ protected
     QQueue<Task::Ptr> m_queue;
 
-    QHash<Task*, Task::Ptr> m_doing;
+    QHash<Task*, Task::Ptr> m_doing; 
```
trailing space

Scrumplex commented on launcher/tasks/Task.cpp:
```diff
@@ -37,8 +38,11 @@
 
 #include <QDebug>
 
+Q_LOGGING_CATEGORY(taskLogC, "launcher.task")
```
hmm?

Scrumplex commented on launcher/tasks/Task.h:
```diff
@@ -36,9 +37,40 @@
 
 #include "QObjectPtr.h"
 
+Q_DECLARE_LOGGING_CATEGORY(taskLogC)
```
hmm?

Ryex committed:
refactor:  move functions to utils + code-review fixes

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
doc fixes

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

TayouVR commented on launcher/qtlogging.ini:
```diff
@@ -0,0 +1,16 @@
+# prevent log spam and strange bugs
+# qt.qpa.drawing in particular causes theme artifacts on MacOS
+qt.*.debug=false
+# dont log credentials buy defualt
```
```suggestion
# don't log credentials by default
```

DioEgizio commented:
Conflicting with develop

Ryex committed:
Merge remote-tracking branch 'upstream/develop' into better-tasks

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
typo fix

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Scrumplex commented on launcher/StringUtils.cpp:
```diff
@@ -113,6 +115,69 @@ int StringUtils::naturalCompare(const QString& s1, const QString& s2, Qt::CaseSe
+}
+
+static const QStringList s_units_si{ "KB", "MB", "GB", "TB" };
+static const QStringList s_units_kibi{ "KiB", "MiB", "Gib", "TiB" };
```
```suggestion
static const QStringList s_units_kibi{ "KiB", "MiB", "GiB", "TiB" };
```

Ryex committed:
Gib -> GiB

Co-authored-by: Sefa Eyeoglu <contact@scrumplex.net>
Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex commented:
Woo! three approvals!
